dry_adiabatic_adjust/dadadj.F90

```diff
@@ -0,0 +1,219 @@
+    real(kind_phys), allocatable :: c3dad(:) ! intermediate constant
+    real(kind_phys), allocatable :: c4dad(:) ! intermediate constant
+    real(kind_phys) :: gammad    ! dry adiabatic lapse rate (deg/Pa)
+    real(kind_phys) :: zeps      ! convergence criterion (deg/Pa)
+    real(kind_phys) :: rdenom    ! reciprocal of denominator of expression
+    real(kind_phys) :: dtdp      ! delta-t/delta-p
+    real(kind_phys) :: zepsdp    ! zeps*delta-p
+    real(kind_phys) :: zgamma    ! intermediate constant
+    real(kind_phys) :: qave      ! mean q between levels
+    real(kind_phys) :: cappaint  ! Kappa at level intefaces
+    real(kind_phys), pointer :: t(:,:)
+    real(kind_phys), pointer :: q(:,:)
+
+    logical :: ilconv          ! .TRUE. ==> convergence was attained
+    logical :: dodad(ncol)     ! .TRUE. ==> do dry adjustment
+
+    !-----------------------------------------------------------------------
+
+    zeps = 2.0e-5_kind_phys           ! set convergence criteria
+    errmsg = ''
+    errflg = 0
+    scheme_name = 'DADADJ'
+
+    allocate(c1dad(nlvdry), stat=errflg)
+    if (errflg /= 0) then
+       errmsg = trim(scheme_name)//': Allocate of c1dad(nlvdry) failed'
+       return
+    end if
+    allocate(c2dad(nlvdry), stat=errflg)
+    if (errflg /= 0) then
+       errmsg = trim(scheme_name)//': Allocate of c2dad(nlvdry) failed'
+       return
+    end if
+    allocate(c3dad(nlvdry), stat=errflg)
+    if (errflg /= 0) then
+       errmsg = trim(scheme_name)//': Allocate of c3dad(nlvdry) failed'
+       return
+    end if
+    allocate(c4dad(nlvdry), stat=errflg)
+    if (errflg /= 0) then
+       errmsg = trim(scheme_name)//': Allocate of c4dad(nlvdry) failed'
+       return
+    end if
+
+    ! t_tend< and tend_dtdq used as workspace until needed to calculate tendencies
+    t => t_tend
+    q => q_tend
+
+    t = state_t
+    q = state_q
+
+    ! Find gridpoints with unstable stratification
+
+    do i = 1, ncol
+       cappaint = 0.5_kind_phys*(cappa(i,2) + cappa(i,1))
+       gammad = cappaint*0.5_kind_phys*(t(i,2) + t(i,1))/pint(i,2)
+       dtdp = (t(i,2) - t(i,1))/(pmid(i,2) - pmid(i,1))
+       dodad(i) = (dtdp + zeps) > gammad
+    end do
+
+    dadpdf(:ncol,:) = 0._kind_phys
+    do k= 2, nlvdry
+       do i = 1, ncol
+         cappaint = 0.5_kind_phys*(cappa(i,k+1) + cappa(i,k))
+         gammad = cappaint*0.5_kind_phys*(t(i,k+1) + t(i,k))/pint(i,k+1)
+         dtdp = (t(i,k+1) - t(i,k))/(pmid(i,k+1) - pmid(i,k))
+         dodad(i) = dodad(i) .or. (dtdp + zeps) > gammad
+         if ((dtdp + zeps) > gammad) then
+            dadpdf(i,k) = 1._kind_phys
+         end if
+      end do
+   end do
+
+   ! Make a dry adiabatic adjustment
+   ! Note: nlvdry ****MUST**** be < pver
+
+   COL: do i = 1, ncol
+
+      if (dodad(i)) then
+
+         zeps = 2.0e-5_kind_phys
+
+         do k = 1, nlvdry
+            cappaint = 0.5_kind_phys*(cappa(i,k+1) + cappa(i,k))
```

Comment:
Was the addition of this line a bug that needed to be fixed? Does it not change answers between this and what is currently in CAM?

Reply:
I think it is a bug, introduced when variable cappa was added to the model.  I've sent this off to Peter to see if he agrees and if he needs to see a control before this is approved.

Reply:
Peter approved the change and doesn't think we need a control.